File: src/main/java/com/example/demo/controller/ExampleController.java

package com.example.demo.controller;

import com.example.demo.dto.ExampleDto;
import com.example.demo.service.ExampleService;
import com.example.demo.validation.OnCreate;
import com.example.demo.validation.OnUpdate;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/examples")
public class ExampleController {

    private final ExampleService service;

    public ExampleController(ExampleService service) {
        this.service = service;
    }

    @PostMapping
    public ResponseEntity<ExampleDto> create(@RequestBody @Validated(OnCreate.class) ExampleDto dto) {
        return ResponseEntity.status(201).body(service.create(dto));
    }

    @PatchMapping("/{id}")
    public ResponseEntity<ExampleDto> update(@PathVariable Long id, @RequestBody @Validated(OnUpdate.class) ExampleDto dto) {
        return ResponseEntity.ok(service.update(id, dto));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ExampleDto> get(@PathVariable Long id) {
        return ResponseEntity.ok(service.get(id));
    }

    @GetMapping
    public ResponseEntity<List<ExampleDto>> getAll() {
        return ResponseEntity.ok(service.getAll());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        service.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/com/example/demo/controller/ExampleController.kt
package com.example.demo.controller

import com.example.demo.dto.ExampleDto
import com.example.demo.service.ExampleService
import com.example.demo.validation.OnCreate
import com.example.demo.validation.OnUpdate
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/examples")
class ExampleController(private val service: ExampleService) {

    @PostMapping
    fun create(@RequestBody @Validated(OnCreate::class) dto: ExampleDto): ResponseEntity<ExampleDto> =
        ResponseEntity.status(201).body(service.create(dto))

    @PatchMapping("/{id}")
    fun update(@PathVariable id: Long, @RequestBody @Validated(OnUpdate::class) dto: ExampleDto): ResponseEntity<ExampleDto> =
        ResponseEntity.ok(service.update(id, dto))

    @GetMapping("/{id}")
    fun get(@PathVariable id: Long): ResponseEntity<ExampleDto> =
        ResponseEntity.ok(service.get(id))

    @GetMapping
    fun getAll(): ResponseEntity<List<ExampleDto>> =
        ResponseEntity.ok(service.getAll())

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): ResponseEntity<Void> {
        service.delete(id)
        return ResponseEntity.noContent().build()
    }
}
